### PR TITLE
perf(megatron): autotune fused LM-head Triton kernels

### PR DESCRIPTION
**Pull request by dyurk-lila**
## Summary

The fused Triton LLM head kernel was originally vendored from VERL with basically no changes. This PR does make two performance-improving changes, without changing any of the underlying computation logic:

- Use triton autotune to optimize block sizes, K tiles, pipeline stages, and warp counts rather than keeping one fixed set for every model/context len/hardware.
- Add an option to exclude entropy calculation entirely if the caller doesn't want to know entropy anyway

Other changes in the PR include:

- Make autotuned epilogues non-mutating. Triton executes every candidate during tuning, so overwriting candidate inputs can silently corrupt the selected result.
- Fix an edge case where the kernels could NaN if all incoming logits were strongly negative
- Remove vendored reductions and backward strategies that no SkyRL path can reach.
- Extend the existing standalone fused-LM-head benchmark and GPU tests.

Note: the LoC seems big, but most deletions are removing unused VERL code and most additions are either in the benchmark script or just adding the triton autotune config lines. The scope of relevant logic changes is relatively small.

## H100 benchmark

The standalone sweep used one 8x NVIDIA H100 80GB node with 180 CPU requested and limited. It covered four public LM-head configurations, total token counts from 16K through 256K, and every `(TP, CP)` combination in `{1,2,4,8}` with `TP*CP <= 8`. The 200 logical cases collapse to 104 unique rank-local `(M,H,V)` shapes. This is a synthetic kernel-shape sweep; it does not consume dataset examples.

Compared with the prior fixed five-stage mainloop:

| Metric across 104 local shapes | Autotuned |
|---|---:|
| Mean latency reduction | 5.57% |
| Median latency reduction | 3.63% |
| p90 latency reduction | 14.47% |
| Maximum latency reduction | 26.82% |
| Shapes faster | 88 / 104 |
| Worst regression | 1.40% |

The largest observed win was the Nemotron Ultra 550B shape at local `(M=8192,H=8192,V=131072)`, from 34.91 ms to 25.54 ms. These are isolated fused-mainloop results on one GPU generation, not end-to-end training claims; the candidate set intentionally retains schedule diversity for other models and accelerators.

## Compile and tuning behavior

A specialization probe measured:

- 20,000 tokens, new 32K bucket: 13.27 s cold compile+tune;
- 20,001 tokens, same bucket: 6.13 ms, with no exact-length recompile or retune;
- 40,000 tokens, new 64K bucket: 4.66 s retune.

Crossing a power-of-two bucket intentionally retunes because the winning geometry can change. Broad cold tuning is nontrivial (14.9 s median first call in this sweep), so long-running jobs should warm expected buckets before timing.

## Correctness

Autotuning originally exposed a real hazard: an epilogue overwrote reductions that later candidates consumed. The final kernels use distinct raw, reduced, and result buffers.

Two 8-GPU H100 gates passed against independently materialized fp32 references:

- every one of the 14 forward candidates matched, with maximum absolute error `9.54e-7`;
- odd/non-divisible shapes, TP=8 reductions, entropy compatibility, and split-N backward matched with maximum errors `1.91e-6` for reduced log-probs/entropy, `2.34e-5` for hidden gradients, and `2.80e-5` for weight gradients.

Additional hardening initializes log-sum-exp from the true row maximum (`-inf`) for strongly negative rows and releases the final partial-split d-logits staging buffer before its projections. A pinned-config audit found bit-identical results before and after the lifetime change across 24 TP/dtype/shape cases.

## Vendored VERL history

VERL used `@triton.autotune`, but each config list contained one entry, so it performed no search. The kernel originated in [VERL PR #462](https://github.com/verl-project/verl/pull/462); its comments, discussion, reviews, and commit messages do not explain the singleton sets. The original benchmark covered one bf16 workload, which may explain the narrow schedule but is not documented upstream rationale.

---

> [!NOTE]
> **Medium Risk**
> Changes sit on the Megatron fused LM-head hot path (forward/backward numerics, TP reductions, and first-call Triton compile/autotune latency); fixes reduce NaN risk but new tuning behavior can affect performance and timing.
> 
> **Overview**
> Reworks the vendored fused LM-head Triton path for **real schedule search** and a **log-prob-only** training path, plus correctness fixes on the softmax epilogue.
> 
> **Autotuning:** Forward mainloop, epilogues (including TP), and split-N backward now use multi-candidate `@triton.autotune` with `cache_results=True`. Cache keys use power-of-two **token buckets** (`num_tokens_bucket`) while launches still use the exact token count (`do_not_specialize`) to limit recompile churn.
> 
> **Entropy optional:** `compute_entropy` gates entropy buffers and kernel work; `FusedLinearLogprobTriton` sets `compute_entropy=False` on forward/backward. Scalar CE reductions and the unused VERL `LinearCrossEntropy` wrapper / alternate backward kernels are removed; backward is only the split-N d-logits path.
> 
> **Correctness / memory:** Epilogues write final log-probs (and entropy) to **separate result buffers** so autotune replays do not corrupt reductions. Log-sum-exp uses **`-inf` init and padding** for max reductions. The last backward split **drops the full-width d-logits staging buffer** before matmul projections.
> 
> **Tooling:** `bench_fused_linear_logprob.py` adds `--autotune-sweep`, entropy-vs-logprob probes, and stricter multi-rank correctness. GPU tests cover bucket math, autotune metadata, strongly negative logits, and adapter `compute_entropy=False`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit cfbaff0afe58cd384674a064cd1613bacd1d1641. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: skyrl/backends/skyrl_train/distributed/megatron/fused_linear_logprob_triton.py [hunk 1]
```diff
@@ -140,93 +147,130 @@ def alloc_fn(size: int, alignment: int, stream: typing.Optional[int]):
...
 
 @triton.autotune(
-    configs=[triton.Config({"BLOCK_SIZE_M": 128, "BLOCK_SIZE_N": 256, "BLOCK_SIZE_K": 32}, num_stages=3, num_warps=8)],
-    key=["num_tokens", "hidden_size", "vocab_size"],
+    configs=_matmul_autotune_configs(_FORWARD_MAINLOOP_CONFIG_SPECS),
+    key=["num_tokens_bucket", "hidden_size", "vocab_size", "COMPUTE_ENTROPY"],
+    cache_results=True,
 )
-@triton.jit
+@triton.jit(do_not_specialize=["num_tokens", "num_tokens_bucket"])
```
[the remark this answers is not in the document]
Comment (dyurk-lila) [reply skyrl/backends/skyrl_train/distributed/megatron/fused_linear_logprob_triton.py]: Addressed in `292672fd`: `rank` is included in `do_not_specialize` for the forward mainloop kernel, avoiding rank-specific compilation.

File: skyrl/backends/skyrl_train/distributed/megatron/fused_linear_logprob_triton.py [hunk 2]
```diff
@@ -776,643 +840,48 @@ def epilogue_grid(meta):
...
-            num_warps=8,
-        ),
-    ],
-    key=["num_tokens", "hidden_size", "vocab_size"],
+    configs=_backward_autotune_configs(),
+    key=["num_tokens_bucket", "hidden_size", "vocab_size", "COMPUTE_ENTROPY"],
+    cache_results=True,
 )
-@triton.jit
+@triton.jit(do_not_specialize=["split_idx", "num_tokens", "num_tokens_bucket"])
```
[the remark this answers is not in the document]
Comment (dyurk-lila) [reply skyrl/backends/skyrl_train/distributed/megatron/fused_linear_logprob_triton.py]: Addressed in `292672fd`: `rank` is included in `do_not_specialize` for the backward split-N kernel as well.

File: skyrl/benchmarks/bench_fused_linear_logprob.py [hunk 3]
```diff
@@ -121,31 +188,381 @@ def _active_modes():
...
+    triton = _require_triton()
+
+    def grid(meta):
+        return (triton.cdiv(shape.num_tokens, meta["BLOCK_SIZE_M"]) * buffers["num_splits"],)
+
+    module.efficient_entropy_kernel_general_mainloop[grid](
+        *_kernel_args(module, shape, buffers),
+        USE_TMA=module.SUPPORT_CUDA_TMA,
+        INPUT_PRECISION="tf32",
+    )
```
[the remark this answers is not in the document]
Comment (dyurk-lila) [reply skyrl/benchmarks/bench_fused_linear_logprob.py:421]: Fixed in `d797c39b`: both benchmark sweep launch paths now pass the required `COMPUTE_ENTROPY` constexpr.

File: skyrl/train/dataset/samplers.py [hunk 4]
```diff
@@ -56,6 +70,236 @@ def import_sampler_class(class_path: str) -> type:
...
+            )
+        if batch_size <= 0 or dp_size <= 0:
+            raise ValueError(f"batch_size and dp_size must be positive, got {batch_size=} and {dp_size=}.")
+        if not 0 < allowed_variation < 1:
+            raise ValueError(f"allowed_variation must be in (0, 1), got {allowed_variation}.")
+
+        self.sampler = sampler
+        self.batch_size = batch_size
+        self.dp_size = dp_size
+        self.max_variation = math.floor(batch_size * allowed_variation)
```
[the remark this answers is not in the document]
Comment (dyurk-lila) [reply skyrl/train/dataset/samplers.py:111]: This rounding is intentional. For a batch smaller than 20, moving one example would exceed the configured 5% cardinality variation, so there is no legal boundary shift. DP bin padding is the expected fallback; users can raise the allowed variation if they want shifts on small batches.

File: skyrl/train/dataset/bin_packing.py [hunk 5]
```diff
@@ -211,8 +233,119 @@ def _pack_implementation(self, sequence_lengths: List[int]) -> List[List[int]]:
...
+                if bin_length + size <= max_unpadded_length:
+                    leftover_bins[bin_index].append(item)
+                    leftover_bin_lengths[bin_index] += size
+                    break
+            else:
+                leftover_bins.append([item])
+                leftover_bin_lengths.append(size)
+        bins.extend(leftover_bins)
+
+        return [[index for index, _ in bin_items] for bin_items in bins if bin_items]
```
[the remark this answers is not in the document]
Comment (dyurk-lila) [reply skyrl/train/dataset/bin_packing.py:343]: This is now the intended MFFD soft-cap behavior for an oversized aligned singleton. Its physical alignment is unavoidable, so the sequence gets its own bin; `_fits` prevents another sequence from joining it over budget. The aligned singleton case has a regression test in #2108.
